src/main.rs

```
@@ -1,3 +1,76 @@
+mod memory;
+mod opdcodes;
+mod registers;
+
+use memory::Memory;
+use opdcodes::Opcode;
+use registers::Registers;
+
+struct VM {
+    memory: Memory,
+    registers: Registers,
+}
+
+impl VM {
+    pub fn new() -> Self {
+        Self {
+            memory: Memory::new(),
+            registers: Registers::new(),
+        }
+    }
+
+    pub fn read_memory(&self, address: u16) -> u16 {
+        self.memory.read(address)
+    }
+
+    pub fn write_memory(&mut self, address: u16, value: u16) {
+        self.memory.write(address, value);
+    }
+
+    pub fn read_register(&self, r: usize) -> u16 {
+        self.registers.get(r).unwrap_or(0)
```

Defaulting to 0 can hide errors and make it harder to debug later, I would return an error